data/transactions/logic/pairing_test.go

```diff
@@ -16,14 +16,876 @@
+			benchmarkOperation(b, g1teal, dups+"dup; extract 0 32;"+dups+"ec_multi_exp BN254g1", "len")
+		})
+	}
+
+	b.Run("g2 scalar_mul", func(b *testing.B) {
+		benchmarkOperation(b, g2teal, "dup; extract 0 32; ec_scalar_mul BN254g2", "len")
+	})
+
+	for i := 0; i < 6; i++ {
+		size := 1 << uint(i)
+		dups := strings.Repeat("dup; concat;", i)
+		b.Run(fmt.Sprintf("g2 multi_exp %d", size), func(b *testing.B) {
+			benchmarkOperation(b, g2teal, dups+"dup; extract 0 32;"+dups+"ec_multi_exp BN254g2", "len")
+		})
+	}
+
+	var g1GenNeg bn254.G1Affine
+	g1GenNeg.Neg(&bnG1Gen)
+	g1points := []bn254.G1Affine{g1GenNeg, bnG1Gen}
+	g2points := []bn254.G2Affine{bnG2Gen, bnG2Gen}
+	// -1 g1 g2 + g1 g2 = 0
+	g1pbytes := tealBytes(bn254G1sToBytes(g1points))
+	g2pbytes := tealBytes(bn254G2sToBytes(g2points))
+
+	b.Run("pairing 1", func(b *testing.B) {
+		benchmarkOperation(b, "", g1teal+g2teal+"ec_pairing_check BN254g1; !; assert", "int 1")
+	})
+	for i := 0; i < 4; i++ {
+		size := 1 << uint(i)
+		dups := strings.Repeat("dup; concat;", i)
+
+		// size * 2 in name because we start with two points
+		b.Run(fmt.Sprintf("pairing %d", size*2), func(b *testing.B) {
+			benchmarkOperation(b, "", g1pbytes+dups+g2pbytes+dups+"ec_pairing_check BN254g1; assert", "int 1")
+		})
+	}
+
+	b.Run("g1 subgroup", func(b *testing.B) {
+		benchmarkOperation(b, "", g1teal+"ec_subgroup_check BN254g1; assert", "int 1")
+	})
+	b.Run("g2 subgroup", func(b *testing.B) {
+		benchmarkOperation(b, "", g2teal+"ec_subgroup_check BN254g2; assert", "int 1")
+	})
+
+	fpbytes := fmt.Sprintf("byte 0x%s\n",
+		strings.Repeat("00", 1)+strings.Repeat("22", bn254fpSize-1))
+	fp2bytes := fpbytes + fpbytes + "concat\n"
+
+	b.Run("g1 map to", func(b *testing.B) {
+		benchmarkOperation(b, "", fpbytes+"ec_map_to BN254g1; pop", "int 1")
+	})
+	b.Run("g2 map to", func(b *testing.B) {
+		benchmarkOperation(b, "", fp2bytes+"ec_map_to BN254g2; pop", "int 1")
+	})
+
+}
+
+func BenchmarkFindMultiExpCutoff(b *testing.B) {
+	for i := 1; i < 10; i++ {
+		kbytes := make([]byte, i*scalarSize)
+		rand.Read(kbytes)
+		{
+			g1points := make([]bls12381.G1Affine, i)
+			b.Run(fmt.Sprintf("bls g1 small %02d", i), func(b *testing.B) {
+				for r := 0; r < b.N; r++ {
+					for j := 0; j < i; j++ {
+						g1points[j] = bls12381RandomG1()
+					}
+					rand.Read(kbytes)
+					bls12381G1MultiExpSmall(g1points, kbytes)
+				}
+			})
+			b.Run(fmt.Sprintf("bls g1 large %02d", i), func(b *testing.B) {
+				for r := 0; r < b.N; r++ {
+					for j := 0; j < i; j++ {
+						g1points[j] = bls12381RandomG1()
+					}
+					rand.Read(kbytes)
+					bls12381G1MultiExpLarge(g1points, kbytes)
+				}
+			})
+
+			g2points := make([]bls12381.G2Affine, i)
+			b.Run(fmt.Sprintf("bls g2 small %02d", i), func(b *testing.B) {
+				for j := 0; j < i; j++ {
+					g2points[j] = bls12381RandomG2()
+				}
+				rand.Read(kbytes)
+				for r := 0; r < b.N; r++ {
+					bls12381G2MultiExpSmall(g2points, kbytes)
```

Comment:
Is it worth calling `b.ResetTimer()` before the loop?

It's also not clear to me why some code includes the `g1points/g2points` setting and `rand.Read(kbytes)` call inside of the loop over `b.N` and some does not.

Reply:
I don't think there's any point in resetting - the only goal of this benchmark is to find out when one function gets slower than the other. Since they both have the randomness at the beginning, they should have the same overhead, and the cutoff will still be accurate.

Except, and I think this goes to your second question, it looks like I forgot to rerandomize the points for G1 large.  I'll add that.  I will also tuck the rerandomization (for all calls) inside the loop, since I've seen some of these function perform better on repeated runs against the same values.